File: Practice Problems/Stacks/Basics of Stacks/A Game of Numbers.cpp
/*
https://www.hackerearth.com/practice/data-structures/stacks/basics-of-stacks/practice-problems/algorithm/a-game-of-numbers-1-5d3a8cb3/
*/

#include <iostream>

using namespace std;

 

int main(){

int n;

cin>>n;

int arr[n]={};

for (int i{};i<n;i++){

cin>>arr[i];

}
 

int greater_index[n]={};

bool done{false};

int smaller[n]={};

 

for (int i{};i<n;i++){

done=false;

for (int j{i};j<n;j++){

if (arr[j]>arr[i]){

greater_index[i]=j;

done=true;

break;

}

}

if (done==false)

greater_index[i]=-1;

}

 

for (int i{};i<n;i++){

done=false;

for (int j{i};j<n;j++){

if (arr[j]<arr[i]){

smaller[i]=arr[j];

done=true;

break;

}

}

if (done==false)

smaller[i]=-1;

}

 

for (int i{};i<n;i++){

if (greater_index[i]==-1)

cout<<-1<<' ';

else

cout<<smaller[greater_index[i]]<<' ';

}
 

}

File: Practice Problems/Stacks/Basics of Stacks/A Game of Numbers.py
"""
https://www.hackerearth.com/practice/data-structures/stacks/basics-of-stacks/practice-problems/algorithm/a-game-of-numbers-1-5d3a8cb3/
"""

n=int(input())
a=[]
for i in range(n):
    x=int(input())
    a.append(x)
st=[]
f=[-1 for i in range(n)]
g=[-1 for i in range(n)]
for i in range(n-1,-1,-1):
    while(st and a[st[-1]]<=a[i]):
        st.pop()
    if  st : f[i]=st[-1]
    st.append(i)
st=[]
#CBGKS
for i in range(n-1,-1,-1):
    while(st and a[st[-1]]>=a[i]):
        st.pop()
    if  st:g[i]=st[-1]
    st.append(i)

l=[-1 for i in range(n)]
for i in range(0,n):
    if(f[i]!=-1 and g[f[i]]!=-1):
         l[i] =a[g[f[i]]]
for i in l:
    print(i,end=" ")
